runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -8433,6 +8442,17 @@ class INTERPRETER_CLASS
 			_sendMethod = *(J9Method **)(((UDATA)receiverClass) + vTableOffset);
 		}
 
+		/* Check if the target is an invokeBasic call and inline the invokeBasic INL to avoid passing
+		 * additional flags during the VM transition
+		 */
+		if (isVararg
+		&& (J9_BCLOOP_SEND_TARGET_METHODHANDLE_INVOKEBASIC == J9_BCLOOP_DECODE_SEND_TARGET(_sendMethod->methodRunAddress))
+		) {
+			j9object_t lambdaForm = J9VMJAVALANGINVOKEMETHODHANDLE_FORM(_currentThread, receiverObject);
+			j9object_t memberName = J9VMJAVALANGINVOKELAMBDAFORM_VMENTRY(_currentThread, lambdaForm);
+			_sendMethod = (J9Method *)(UDATA)J9OBJECT_U64_LOAD(_currentThread, memberName, _vm->vmtargetOffset);
```

Comment:
I'm having trouble saying this is ok or not because I'm not clear on why the method being varargs matters when calling `invokeBasic`.  Is it related to `invokeBasic` being signaturePolymorphic?

Since this is for varargs, is there a need to collect the varargs into the trailing array here?

Sorry, I'm not clear why this code needs to handle varargs specially when the implementation of the `invokeBasic` VM_BytecodeAction doesn't.

Reply:
>Is it related to invokeBasic being signaturePolymorphic?

Yes, I guess I should have used a more exact check instead of `J9AccVarArgs` to avoid including regular varargs method which doesn't require the extra `methodTypeParameterSlotCount` call.

The VM_BytecodeAction `invokeBasic` impl reads the methodArgCount from ramCP which is based on the call signature in CP hence it knows where to locate the receiver object on the stack to load the target `MemberName` from

With methods called by `linkToVirtual`, since transition to new target is still on the same bytecode
(ie. `invokeStatic MethodHandle.linkToVirtual(...)`) the ramCP's methodArgCount would hold the number of parameters passed to `linkToVirtual` which doesn't match what target method actually received. So when looking for the receiver object, we used the `romMethod->ArgCount`

But because `invokeBasic` have a `romMethod->argCount` that may not be consistent with the call signature. the code have to find the actual number of argument expected through the signature from target MemberName's methodType.

Reply:
So this is the case where we have a `linkToVirtual` redispatching to `invokeBasic`?  

And I assume this is created by something like `MethodHandles.lookup().findVirtual(MethodHandle. invokeBasic, MethodType.methodType(...))`?  If that's the case, shouldn't we recognize that pattern on creation and return the `invokeBasic` directly rather than the `linkToVirtual` handle?

In our original implementation, we had special MH subclass to cover the signaturePolymorphic methods called `InvokeExactHandle` & `InvokeGenericHandle`.  I'd expect the OpenJDK impl to shortcut these cases in the resolve of the memberName as well.

Reply:
this is generated when the MH invoke arity is 254, since max JVM argCount is 255, using a invokeHandle instruction would not be possible as it adds extra MemberName and Appendix object to the argument list.

Reply:
So we're dealing with an extreme corner case here?  All that's critical info to put into the source code so we can understand the **why** behind the design.  If you include that info in a comment up front we don't have to try and reverse engineer it when reviewing the change.... and when you have to fix a bug in 6 months you'll be happy to have the info directly in the code.

Can you update this with a comment that describes that this is an uncommon case and why it has to be handled this way?

Reply:
I will add a comment to describe this special case and the reasons for such handling

Reply:
Can you also add a comment to the `invokeBasic` bytecodeAction to indicate that the code has been inlined here?  It'll make it easier to keep both copies in sync if they refer to each other